### PR TITLE
expression: implement castTimeAsDuration function pushdown

### LINKED ISSUE #37033: Implement castTimeAsDuration function push down to TiFlash
## Enhancement
support CastTimeAsDuration in TiFlash
For details: [https://github.com/pingcap/tiflash/pull/5548](https://github.com/pingcap/tiflash/pull/5548)

### PR DESCRIPTION
**Pull request by AntiTopQuark**
### What problem does this PR solve?

Issue Number: close #37033
related issue: https://github.com/pingcap/tiflash/issues/5306
Problem Summary:
support castTimeAsDuration in tiflash

### What is changed and how it works?
1. open pushdown of  castTimeAsDuration
2. add some test

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support to pushdown castTimeAsDuration to TiFlash
```

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -7091,3 +7091,40 @@ func TestEltPushDownToTiFlash(t *testing.T) {
...
+		}
+	}
+
+	rows := [][]interface{}{
+		{"TableReader_9", "root", "data:ExchangeSender_8"},
+		{"└─ExchangeSender_8", "mpp[tiflash]", "ExchangeType: PassThrough"},
+		{"  └─Projection_4", "mpp[tiflash]", "cast(test.t.b, time BINARY)->Column#4"},
+		{"    └─TableFullScan_7", "mpp[tiflash]", "keep order:false, stats:pseudo"},
+	}
+	tk.MustQuery("explain select cast(b as time) from t;").CheckAt([]int{0, 2, 4}, rows)
```
Comment (SeaRise) [inline planner/core/integration_test.go:7129]:
```suggestion
	tk.MustQuery("explain select cast(a as time), cast(b as time) from t;").CheckAt([]int{0, 2, 4}, rows)
```

File: expression/expression.go [hunk 2]
```diff
@@ -1105,6 +1105,8 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
 			tipb.ScalarFuncSig_CastStringAsTime /*, tipb.ScalarFuncSig_CastDurationAsTime, tipb.ScalarFuncSig_CastJsonAsTime*/ :
 			// ban the function of casting year type as time type pushing down to tiflash because of https://github.com/pingcap/tidb/issues/26215
 			return function.GetArgs()[0].GetType().GetType() != mysql.TypeYear
+		case tipb.ScalarFuncSig_CastTimeAsDuration:
+			return sourceType.GetType() == retType.GetType() || retType.GetType() == mysql.TypeDuration
```
Comment (SeaRise) [inline expression/expression.go:1109]: why `sourceType.GetType() == retType.GetType()`?

Comment (AntiTopQuark) [reply expression/expression.go]: Sorry
seeing the above code, I thought I had to write it this way
I have modified it

**Review by SeaRise — APPROVED** [review]
LGTM

**Review by ywqzzy — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

File: expression/expr_to_pb_test.go [hunk 3]
```diff
@@ -1185,6 +1185,11 @@ func TestExprPushDownToFlash(t *testing.T) {
 	pushed, remained = PushDownExprsWithExtraInfo(sc, exprs, client, kv.TiFlash, true)
 	require.Len(t, pushed, len(exprs))
 	require.Len(t, remained, 0)
+
+	// CastTimeAsDuration
+	function, err = NewFunction(mock.NewContext(), ast.Cast, types.NewFieldType(mysql.TypeDatetime), durationColumn)
+	require.NoError(t, err)
+	exprs = append(exprs, function)
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply expression/expr_to_pb_test.go:1192]: yes...

[the remark this answers is not in the document]
Comment (SeaRise) [reply expression/expr_to_pb_test.go:1192]: https://github.com/pingcap/tidb/pull/37115

**Pull request merged**